### DISCUSSION TITLE
why asp.net core doesn't support AOP?

### DISCUSSION THREAD

**Discussion by diymine** [dotnet/aspnetcore#23515] [General]
Hi All,

I have greater passion to study asp.net core mvc but actually most companies like to use java spring to set up their new projects.
Now I am a member of java.

When I study java, I am confused why asp.net core doesn't default support AOP?
AOP is supported by spring mvc every where and very deep.

I knew asp.net core can support AOP by extends autofac, but it seems that it doesn't support the case with async/await

and for this basic function, I thought it need to be support by asp.net core framework default or intergration easily but actually not

So could you give me any reason or roadmap?

Thanks

**Comment by davidfowl** [comment]
The framework has specific extensibility points but there's no global AOP support in ASP.NET Core itself. It's not something we plan to add either. That can be implemented in various ways:
- Autofac as you mention
- [Castle.Core](https://www.nuget.org/packages/Castle.Core/)
- [Fody](https://github.com/Fody/Fody)

**Reply by diymine** [reply]
As I said, the Autofac doesn't support the async/await method.
It seems that Autofac use the Castle.Core to do the AOP(intercept),
the Castle.Core provide a choice that using [Castle.Core.AsyncInterceptor](https://github.com/JSkimming/Castle.Core.AsyncInterceptor) to intercept async/await method, but seems not very easy. I saw in the issue discussion that provide a bridge method to implement it but not got time to try it.
Not sure if the Fody support it.

Maybe c# is more advance with simple asynchronous feature(async/await I like very much) that makes the AOP more complex

But async/await is an advanced feature that will be used by many companies/projects/new users
and it is hard to intergrate it with AOP.
Also asp.net core framework doesn't support it default like java spring mvc and need coder take much time on it before they focus on the business logic if they want to use the AOP with asynchronous methods, it seems it is not a good thing and hope to get any guide from the official 

Thanks

**Reply by davidfowl** [reply]
No problem. Curious, what are you trying to AOP? Everything?

**Reply by diymine** [reply]
Thanks your quick reply.

what I tried to AOP is database transaction intercepter
class TransactionInterceptor : IAsyncInterceptor(the interface in Castle.Core.AsyncInterceptor)
an TransactionAttribute:Attribute to activate the interceptor

In Startup, register the intercepor with autofac like below
ContainerBuilder builder.RegisterType<TransactionInterceptor>() 

But it thow below exception
Unable to cast object of type 'TransactionInterceptor' to type 'Castle.DynamicProxy.IInterceptor'.

Maybe something I make wrong due to poor knowledge.

You can give a stardard example that doesn't need to intercepter the database transaction but the method is an async/await method and the Castle.Core.AsyncInterceptor is not necessary

what I saw most in articles is tried to intercepter log

Thanks

**Reply by davidfowl** [reply]
I was just trying to get a feel for what kinds of things you were trying to intercept.

**Comment by hwangzhihuan** [comment]
他就是不支持。
